File: platforms/desktop/src/commands/project.rs
use sona_core::project::{ProjectCreateInput, ProjectRecord, ProjectUpdateInput};
use tauri::{AppHandle, Runtime};

#[tauri::command]
pub async fn project_list<R: Runtime>(app: AppHandle<R>) -> Result<Vec<ProjectRecord>, String> {
    crate::platform::tag_repository::list_projects(&app).await
}

#[tauri::command]
pub async fn project_create<R: Runtime>(app: AppHandle<R>, input: ProjectCreateInput) -> Result<ProjectRecord, String> {
    crate::platform::tag_repository::create_project(&app, input).await
}

#[tauri::command]
pub async fn project_update<R: Runtime>(app: AppHandle<R>, project_id: String, updates: ProjectUpdateInput) -> Result<Option<ProjectRecord>, String> {
    crate::platform::tag_repository::update_project(&app, project_id, updates).await
}

#[tauri::command]
pub async fn project_delete<R: Runtime>(app: AppHandle<R>, project_id: String, cascade_action: Option<String>) -> Result<(), String> {
    crate::platform::tag_repository::delete_project_with_cascade(&app, project_id, cascade_action.unwrap_or_else(|| "moveToInbox".into())).await
}

#[tauri::command]
pub async fn project_reorder<R: Runtime>(app: AppHandle<R>, project_ids: Vec<String>) -> Result<Vec<ProjectRecord>, String> {
    crate::platform::tag_repository::reorder_projects(&app, project_ids).await
}

#[tauri::command]
pub async fn project_get_active_id<R: Runtime>(app: AppHandle<R>) -> Result<Option<String>, String> {
    crate::platform::tag_repository::get_active_tag_id(&app).await
}

#[tauri::command]
pub async fn project_set_active_id<R: Runtime>(app: AppHandle<R>, project_id: Option<String>) -> Result<(), String> {
    crate::platform::tag_repository::set_active_tag_id(&app, project_id).await
}

File: platforms/desktop/frontend/src/services/tauri/project.ts
import type { ProjectRecord as GeneratedProjectRecord } from '../../bindings';
import {
  DEFAULT_PROJECT_PIPELINE,
  type ProjectCreateInput,
  type ProjectPipelineConfig,
  type ProjectRecord,
  type ProjectUpdateInput,
} from '../../types/project';
import { TauriCommand } from './commands';
import type { TauriCommandArgs } from './contracts';
import { invokeTauri } from './invoke';

export type ProjectListRequest = TauriCommandArgs<typeof TauriCommand.project.list>;

function normalizeProject(record: GeneratedProjectRecord): ProjectRecord {
  const pipeline: ProjectPipelineConfig | undefined = record.pipeline
    ? {
        ...DEFAULT_PROJECT_PIPELINE,
        enabled: record.pipeline.enabled ?? false,
        autoPolish: record.pipeline.autoPolish ?? false,
        polishPresetId: record.pipeline.polishPresetId ?? undefined,
        polishPromptOverride: record.pipeline.polishPromptOverride ?? undefined,
        autoTranslate: record.pipeline.autoTranslate ?? false,
        targetLanguage: record.pipeline.targetLanguage ?? undefined,
        autoSummary: record.pipeline.autoSummary ?? false,
        summaryTemplateId: record.pipeline.summaryTemplateId ?? undefined,
        hotwordSetIds: record.pipeline.hotwordSetIds ?? [],
        replacementSetIds: record.pipeline.replacementSetIds ?? [],
        customTerms: record.pipeline.customTerms ?? [],
        autoExport: record.pipeline.autoExport ?? false,
        exportFormat: record.pipeline.exportFormat as ProjectPipelineConfig['exportFormat'],
        exportDirectory: record.pipeline.exportDirectory ?? undefined,
        exportFileNamePrefix: record.pipeline.exportFileNamePrefix ?? undefined,
      }
    : undefined;
  return {
    id: record.id,
    name: record.name,
    description: record.description ?? '',
    icon: record.icon ?? '',
    color: record.color ?? undefined,
    sortOrder: record.sortOrder,
    createdAt: record.createdAt,
    updatedAt: record.updatedAt,
    pipeline,
  };
}

export async function projectList(_request?: ProjectListRequest): Promise<ProjectRecord[]> {
  void _request;
  const records = await invokeTauri(TauriCommand.project.list);
  return records.map(normalizeProject);
}

export async function projectCreate(input: ProjectCreateInput): Promise<ProjectRecord> {
  return normalizeProject(await invokeTauri(TauriCommand.project.create, { input }));
}

export async function projectUpdate(
  projectId: string,
  updates: ProjectUpdateInput
): Promise<ProjectRecord | null> {
  const record = await invokeTauri(TauriCommand.project.update, { projectId, updates });
  return record ? normalizeProject(record) : null;
}

export async function projectDelete(
  projectId: string,
  cascadeAction: 'moveToInbox' | 'deleteItems' = 'moveToInbox'
): Promise<void> {
  await invokeTauri(TauriCommand.project.delete, { projectId, cascadeAction });
}

export async function projectReorder(projectIds: string[]): Promise<ProjectRecord[]> {
  const records = await invokeTauri(TauriCommand.project.reorder, { projectIds });
  return records.map(normalizeProject);
}

export async function projectGetActiveId(): Promise<string | null> {
  return await invokeTauri(TauriCommand.project.getActiveId);
}

export async function projectSetActiveId(projectId: string | null): Promise<void> {
  await invokeTauri(TauriCommand.project.setActiveId, { projectId });
}
